src/cast_core/src/actors/CrudActor.cpp

```diff
@@ -0,0 +1,615 @@
+            node["BypassDocumentValidation"] = *bypassDocValidation;
+        }
+        auto isOrdered = rhs.ordered();
+        node["Ordered"] = isOrdered;
+        return node;
+    }
+
+    static bool decode(const Node& node, BulkWriteOptions& rhs) {
+        if (!node.IsMap()) {
+            return false;
+        }
+        if (node["BypassDocumentValidation"]) {
+            auto bypassDocValidation = node["BypassDocumentValidation"].as<bool>();
+            rhs.bypass_document_validation(bypassDocValidation);
+        }
+        if (node["Ordered"]) {
+            auto isOrdered = node["Ordered"].as<bool>();
+            rhs.ordered(isOrdered);
+        }
+        if (node["WriteConcern"]) {
+            auto wc = node["WriteConcern"].as<mongocxx::write_concern>();
+            rhs.write_concern(wc);
+        }
+        return true;
+    }
+};
+
+template <>
+struct convert<mongocxx::options::count> {
+    using CountOptions = mongocxx::options::count;
+    static Node encode(const CountOptions& rhs) {
+        Node node;
+        auto h = rhs.hint();
+        if (h) {
+            auto hint = h->to_value().get_utf8().value;
+            node["Hint"] = std::string(hint);
+        }
+        auto limit = rhs.limit();
+        if (limit) {
+            node["Limit"] = *limit;
+        }
+        auto maxTime = rhs.max_time();
+        if (maxTime) {
+            auto maxTimeSpec = genny::TimeSpec(*maxTime);
+            node["MaxTimeMillis"] = maxTimeSpec.count();
+        }
+        return node;
+    }
+
+    static bool decode(const Node& node, CountOptions& rhs) {
+        if (!node.IsMap()) {
+            return false;
+        }
+        if (node["Hint"]) {
+            auto h = node["Hint"].as<std::string>();
+            auto hint = mongocxx::hint(h);
+            rhs.hint(hint);
+        }
+        if (node["Limit"]) {
+            auto limit = node["Limit"].as<int>();
+            rhs.limit(limit);
+        }
+        if (node["MaxTimeMillis"]) {
+            auto maxTime = node["MaxTimeMillis"].as<int>();
+            rhs.max_time(std::chrono::milliseconds(maxTime));
+        }
+        if (node["ReadPreference"]) {
+            auto readPref = node["ReadPreference"].as<mongocxx::read_preference>();
+            rhs.read_preference(readPref);
+        }
+        return true;
+    }
+};
+}  // namespace YAML
+
+namespace genny::actor {
+
+namespace {
+
+enum class StageType {
+    Bucket,
+    Count,
+};
+
+std::unordered_map<std::string, StageType> stringToStageType = {{"bucket", StageType::Bucket},
+                                                                {"count", StageType::Count}};
+
+struct BaseOperation {
+    virtual void run(const mongocxx::client_session& session) = 0;
+    virtual ~BaseOperation() = default;
+};
+
+using OpCallback = std::function<std::unique_ptr<BaseOperation>(
+    YAML::Node, bool, mongocxx::collection, genny::DefaultRandom&)>;
+
+struct AggregateOperation : public BaseOperation {
```

Comment:
I'm dubious of using [`mongocxx::pipeline`](http://mongocxx.org/api/current/classmongocxx_1_1pipeline.html) to run an aggregation pipeline because it doesn't have an escape hatch to run an aggregation stage that isn't known to the C++ driver. This seems like something that would limit the ability for the Query team to use genny to write performance tests for new aggregation stages. My preference would be to do away with the `AggregateOperation` class for the time being and revisit it in a separate pull request. CC @rtimmons

Reply:
ya I would vote for that too